sql/moz-fx-data-shared-prod/udf_js/bootstrap_percentile_ci/udf.sql

```diff
@@ -0,0 +1,243 @@
+/*
+Calculates a confidence interval using an efficient bootstrap sampling technique for a
+given percentile of a histogram.
+
+See:
+- https://mozilla-hub.atlassian.net/browse/DS-2128
+- https://engineering.atspotify.com/2022/03/comparing-quantiles-at-scale-in-online-a-b-testing/
+
+Users must specify an array of percentiles of interest as the first parameter and a
+histogram struct as the second.
+*/
+
+CREATE OR REPLACE FUNCTION udf_js.bootstrap_percentile_ci(
+  percentiles ARRAY<INT64>,
+  histogram STRUCT<values ARRAY<STRUCT<key FLOAT64, value FLOAT64>>>,
+  metric STRING
+)
+RETURNS ARRAY<STRUCT<
+    metric STRING,
+    statistic STRING,
+    point FLOAT64,
+    lower FLOAT64,
+    upper FLOAT64,
+    parameter STRING
+>> DETERMINISTIC
+LANGUAGE js OPTIONS ( library=['gs://moz-fx-data-circleci-tests-bigquery-etl/bq_udf_library/qbinom.js'])
+AS
+"""
+  function histogramSort(histogram) {
+    histogram.sort(function(a, b) {return parseFloat(a.key) - parseFloat(b.key);});
+  }
+
+  function validatePercentile(percentile) {
+    if (percentile < 0 || percentile > 100) {
+      throw "percentile must be a value between 0 and 100";
+    }
+  }
+  
+  function getQuantileSamples(percentile, histogram, n_samples) {
+    /*
+    Returns a sorted array of quantile samples from a `histogram` at a target `percentile`.
+    */
+    validatePercentile(percentile);
+
+    /* ensure the histogram is sorted */
+    histogramSort(histogram);
+
+    /* calculate the cumulative number of elements in the histogram */
+    var total = 0;
+    for (var bin of histogram) {
+        total += bin.value;
+        bin.cumulative_sum = total;
+    };
+
+    var samples = [];
+    /* sample the quantile of interest */
+    for (var i = 0; i < n_samples; i++) {
+      /* draw a random variable from the binomial distribution */
+      var x = qbinom(Math.random(), total, parseFloat(percentile) / 100);
+      if (x in samples) {
+        samples[x].value += 1;
+      } else {
+        /* find the histogram bin associated with the sample */
+        for (const bin of histogram) {
+          if (bin.cumulative_sum >= x) {
+            samples[x] = {'key': bin.key, 'value': 1};
+            break;
+          }
+        };
+      }
+    }
+
+    /* ensure the histogram is sorted */
+    histogramSort(samples);
+
+    /* expand the output */
+    var output = samples.map(bin => Array(bin.value).fill(parseFloat(bin.key))).reduce((prev, curr) => prev.concat(curr));
```

Comment:
I'm getting `Reduce of empty array with no initial value`. Does this happen for bins that are 0?

Reply:
something like 
```
if (arr.length > 0) {
      var output = arr.reduce((prev, curr) => prev.concat(curr));
    } else {
      var output = [];
    }
```
does cause it to run successfully, but I'm not sure if the results are still correct with this?

Reply:
Thanks, great catch! I think you're right, this is caused when all of the histogram bins contain 0 counts (e.g. `[{'key': 1, 'value': 0, 'key': 2, 'value': 0}]`). I am now explicitly checking for that case with `histogramIsNonZero`. I added a test for this case, which failed under the previous version and is passing now.

For these cases, the UDF will now return `NULL` for `lower` / `point` / `upper`.